# .01 Mixed Type Data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import data sets

path = r'C:\Users\jeong\OneDrive\Desktop\CF Files\Achievement 4\07-2023 Instacart Basket Analysis'

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [3]:
# Create a dataframe

df_test = pd.DataFrame()

In [4]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1 , True]

In [5]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [6]:
# Check whether dataframe contains any mixed-type columns

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [7]:
# Fix mixed-type column

df_test['mix'] = df_test['mix'].astype('str')

# .02 Missing Values

In [8]:
# Find total missing observations in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
# Create subset of product_name column and its missing values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
df_prods.shape

(49693, 5)

In [12]:
# Create dataframe without missing values

# df_prods.dropna(inplace = True)

# df_prods.dropna(subset = ['product_name'], inplace = True)

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
df_prods_clean.shape

(49677, 5)

# .03 Duplicates

In [14]:
# Create sbuset of df_prods_clean containing only rows that are duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [15]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
# Number of rows in df_prods_clean without duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exercise 4.5 Step 2

In [19]:
# Summary statistics of df_ords

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_last_order  
count            3.421083e+06       3.421083e+06           3.214874e+06  
mean             2.776219e+00       1.345202e+01           1.111484e+01  
std              2.046829e+00       4.226088e+00           9.206737e+00  
min              0.000000e+00       0.000000e+00           0.000000e+00  
25%              1.000000e+00       1.000000e+01           4.000000e+00  
50%              3.000000e+00       1.300000e+01           7.000000e+00  
75%              5.000000e+00       1.600000e+01           1.500000e+01  
max              6.000000e+00       2.300000e+01           3.000000e+01

In [20]:
df_ords = df_ords.drop(columns = 'Unnamed: 0')

In [21]:
# Summary statistics of df_prods

df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Observations: Max value for 'prices' seem high

In [22]:
# Find record with max 'prices' value

df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [23]:
# Check if there any other incorrect prices

df_prods_clean_no_dups.sort_values('prices', ascending = False)

product_id                      product_name  aisle_id  department_id  \
33666       33664             2 % Reduced Fat  Milk        84             16   
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
19392       19391         Turkey Breast Tenderloins        49             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   
40490       40486                   Chicken Tenders        49             12   
...           ...                               ...       ...            ...   
11307       11307                  Goat Cheese Logs         2             16   
32805       32803                 Regular Deodorant        80             11   
46627       46623                   Pie Pans, Large        10             17   
37041       37037                       Redseedless        24              4   
21464       21463      Vanilla Bean Light Ice Cream        37              1   

        prices  
33666  99999.0  
21554  14900.0  
19392     25.0  
25580     25.0  
40490     25.0  
...        ...  
11307      1.0  
32805      1.0  
46627      1.0  
37041      1.0  
21464      1.0  

[49672 rows x 5 columns]

# Step 3


In [24]:
# Check for mixed-type data in df_ords

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)
    else:
        print('not mixed')

not mixed


# Step 4

In [25]:
# Check if there are missing value in df_ords

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_last_order     206209
dtype: int64

# Step 5

In [26]:
# Create subset for missing values

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [27]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_the_week  \
0         2539329        1             1                       2   
11        2168274        2             1                       2   
26        1374495        3             1                       1   
39        3343014        4             1                       6   
45        2717275        5             1                       3   
...           ...      ...           ...                     ...   
3420930    969311   206205             1                       4   
3420934   3189322   206206             1                       3   
3421002   2166133   206207             1                       6   
3421019   2227043   206208             1                       1   
3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

There are 206209 missing values in the days_since_last_order column.
The NaN seem to be there because these are the first orders for the user_id as we can see by the order_number. 
Therefore, they will not have any previous orders.

# Step 6

The appropiate method in this case would be to leave the missing values as is. Changing the missing value or removing it completely would be altering the first order for each user_id.

# Step 7

In [28]:
# Check and create a subset for duplicate values in df_ords

df_ords_dups = df_ords[df_ords.duplicated()]

In [29]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

There are no duplicates to be found

# Step 8

No action was taken as no duplicates were found

In [30]:
# Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))


In [31]:
# Export df_ords 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))